### PR TITLE
feat: Document tagging system

### LINKED ISSUE #765: Document tags
I think tags would be a really useful addition to outline - is this something you have planned, or an idea how you want to implement it? I could potentially help getting it done.

One method that I think works quite nicely is how Bear (http://bear-writer.com) implements them - you just include a hashtag in a document and it counts as a tag. I've thrown together a very quick and hacky way to implement them in outline using search:

https://github.com/outline/outline/compare/master...mixcloud:hashtags

It uses the fact that postgres full text search thinks that `//tag/...` is a file and doesn't split it when it's indexing. Obviously this is not ideal! Ideally the tags would be clickable in the documents, and I would have created a new migration (since this is a hack I just wanted to quickly try it out without having to learn sequelize though).

### PR DESCRIPTION
**Pull request by dmorel**
## Summary

Implements a full document tagging feature for Outline, as requested in #765.

- **Database**: Two new tables — `tags` (name + team scope, unique index) and `document_tags` (join table with cascade deletes)
- **API**: Six new endpoints — `tags.create` (upsert by name), `tags.list` (with document counts), `tags.update`, `tags.delete` (admin-only), `tags.add`, `tags.remove`
- **Search**: `#tag` tokens in the search bar are parsed client-side, resolved to tag IDs, and passed as `tagIds` filters to `documents.search` and `documents.search_titles` (AND semantics)
- **UI**: TagInput component with typeahead autocomplete on document pages, dismissible tag pills, `/tags` index page in sidebar, tag cards link to search

## Details

- Tags are normalized to lowercase; create is idempotent (upsert)
- Authorization: members can create/update tags and tag documents, only admins can delete tags, viewers are blocked
- Team isolation enforced on all endpoints — cross-team tag operations return 403
- `documents.info` eagerly loads tags; list endpoints skip tags to avoid N+1
- Webhook integration: tag events added to the exhaustive switch-case (currently ignored, ready for future webhook support)
- SearchHelper uses parameterized `EXISTS` subqueries with `sequelize.escape()` for tag filtering

## Test plan

- [x] 18 server API tests covering CRUD, authorization, team isolation, cascade delete, idempotency, documentCount accuracy
- [x] 9 parseSearchQuery unit tests (extraction, normalization, deduplication, edge cases)
- [x] 6 TagsStore unit tests (name cache, case-insensitive lookup, clear, deduplication)
- [x] 1 documents.search tagIds filter integration test
- [x] Full test suite: 2554 passed, 0 regressions (2 pre-existing failures on main)
- [x] TypeScript: `yarn tsc --noEmit` passes with 0 errors
- [x] Lint: 0 new warnings or errors
- [x] Browser-tested: tag CRUD, autocomplete, search with `#tag` tokens, persistence across reload

Closes #765

### CODE REVIEW

File: app/components/TagInput.tsx [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+      }
+
+      if (e.key === "Enter" && inputValue.trim()) {
+        e.preventDefault();
+        const selected =
+          highlightedIndex >= 0
+            ? suggestions[highlightedIndex]
+            : suggestions.length > 0
+              ? suggestions[0]
+              : inputValue;
```
[the remark this answers is not in the document]
Comment (dmorel) [reply app/components/TagInput.tsx]: Fixed in cfaa85b. ArrowDown now includes the Create row in its index range (up to `suggestions.length`), ArrowUp navigates back from it, and Enter prefers the typed value unless an existing suggestion was explicitly highlighted.

File: server/routes/api/tags/tags.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+              `(SELECT COUNT(*) FROM document_tags WHERE document_tags."tagId" = "tag"."id")`
+            ),
+            "documentCount",
+          ],
+        ],
+      },
+      order: [[Sequelize.literal('"documentCount"'), "DESC"]],
+      offset: ctx.state.pagination.offset,
+      limit: ctx.state.pagination.limit,
+    });
```
[the remark this answers is not in the document]
Comment (dmorel) [reply server/routes/api/tags/tags.ts:155]: Not implementing this. The query is paginated (bounded result set per page) and PostgreSQL's planner efficiently handles the correlated subquery with an index on `document_tags(tagId)`. The codebase has no existing precedent for `findAll` with `fn+include+group` aggregation, and mixing GROUP BY with LIMIT/OFFSET and `subQuery: false` in Sequelize has known edge cases. The theoretical benefit doesn't justify the added complexity and risk here.

File: server/routes/api/tags/tags.ts [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+              `(SELECT COUNT(*) FROM document_tags JOIN documents ON documents.id = document_tags."documentId" WHERE document_tags."tagId" = "tag"."id" AND documents."deletedAt" IS NULL)`
+            ),
+            "documentCount",
+          ],
+        ],
+      },
+      order: [[Sequelize.literal('"documentCount"'), "DESC"]],
+      offset: ctx.state.pagination.offset,
+      limit: ctx.state.pagination.limit,
+    });
```
[the remark this answers is not in the document]
Comment (dmorel) [reply server/routes/api/tags/tags.ts:155]: Not implementing this. The query is paginated (bounded result set per page) and PostgreSQL's planner efficiently handles the correlated subquery with an index on `document_tags(tagId)`. The codebase has no existing precedent for `findAll` with `fn+include+group` aggregation, and mixing GROUP BY with LIMIT/OFFSET and `subQuery: false` in Sequelize has known edge cases. The theoretical benefit doesn't justify the added complexity and risk here.

**Comment by pro-ts** [comment]
Basically the most important (at least for me) feature for Outline right now. I hope it will be merged to main soon!

**Comment by Howardzi-nn** [comment]
image

**Comment by fumbles** [comment]
Any progress on deciding to merge these?
Would be a great feature

**Comment by CharlesS42** [comment]
Looking forward to seeing this feature live !

**Comment by swinis** [comment]
Take heart and merge 👍🏻
